[ADD] project_internal_access_from_portal: new module

Mohajiro opened the pull request:
Added a new 'portal_internal' option to the privacy_visibility field in the project.project model, allowing portal users (Portal group) to read projects and tasks marked with this visibility setting.

Added corresponding record rules in security/portal_project_rules.xml and demo data in demo/demo_data.xml for testing.

Task: 4363

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added an option in project settings to allow portal users to access internal projects and tasks.
  - Portal users can now view configured internal projects and their tasks directly from the portal interface.
- **Documentation**
  - Introduced comprehensive documentation, including README, configuration, usage instructions, context, contributor credits, and detailed HTML module description.
- **Demo Data**
  - Added demo project, task, and portal user records to illustrate the new portal access functionality.
- **Security**
  - Implemented access rules to grant portal users read-only access to specified internal projects and tasks.
- **Tests**
  - Added tests to verify portal user access rights and restrictions for internal projects and tasks.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

lbarry-apsl committed:
[16.0][ADD] project_milestone_status: New module project_milestone_status

Mohajiro committed:
[ADD] project_internal_access_from_portal: new module

Add a new 'portal_internal' selection to the privacy_visibility field in
project.project, so portal users can read projects and tasks marked with
this setting.

Added record rules in security/portal_project_rules.xml and demo data in
demo/demo_data.xml for testing.

Task: 4363

Mohajiro commented on project_internal_access_from_portal/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1 @@
+This module adds an additional option to the project settings which allows portal users to access internal projects and tasks.
```
I'll keep it like this, thanks

Mohajiro commented on project_internal_access_from_portal/readme/CONTEXT.md:
```diff
@@ -0,0 +1 @@
+Sometimes you need to provide access to a project to portal users. Even if this project privacy is set to the "Invited internal users"
```
It's okey for me

Mohajiro commented on project_internal_access_from_portal/readme/CONTRIBUTORS.md:
```diff
@@ -0,0 +1,3 @@
+Cetmix<cetmix.com>:
+    Ivan Sokolov
+    Andrei Loukachov
```
Thanks

Mohajiro commented on project_internal_access_from_portal/readme/CONFIGURE.md:
```diff
@@ -0,0 +1,4 @@
+Go to "Project > Configuration > Projects" and open a project.
+In the "Settings" tab set "Visibility" to the "Invited internal/portal users".
```
It's okey

Mohajiro commented on project_internal_access_from_portal/readme/USAGE.md:
```diff
@@ -0,0 +1,2 @@
+When a portal user a configured project. The user can now access the project in the portal.
+When a portal user a configured project task. The user can now access the task in the portal.
```
It's okey too

Mohajiro replied:
It's okey

Mohajiro commented on project_internal_access_from_portal/__manifest__.py:
```diff
@@ -0,0 +1,23 @@
+    """,
+    'category': 'Project',
+    'author': 'Your Company / OCA',
+    'license': 'AGPL-3',
```
Done, thanks

Mohajiro commented on project_internal_access_from_portal/__manifest__.py:
```diff
@@ -0,0 +1,23 @@
+    'version': '16.0.1.0.0',
+    'summary': 'Show internal projects in portal',
+    'description': """
+This module allows portal users to access projects and tasks
+marked as "Invited internal users" by enabling a special flag.
+    """,
```
Deleted

Mohajiro commented on project_internal_access_from_portal/README.rst:
```diff
@@ -0,0 +1,96 @@
+Usage
+=====
+
+When a portal user a configured project. The user can now access the
+project in the portal. When a portal user a configured project task. The
+user can now access the task in the portal.
```
It's okey too

Mohajiro commented on project_internal_access_from_portal/__manifest__.py:
```diff
@@ -0,0 +1,20 @@
+# Copyright (C) 2025 Cetmix OÜ
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+{
+    "name": "Internal Project Available in Portal",
+    "version": "16.0.1.0.0",
+    "summary": "Show internal projects in portal",
+    "website": "https://github.com/OCA/project",
+    "category": "Project",
+    "author": "Odoo Community Association (OCA)",
+    "license": "AGPL-3",
+    "depends": ["project", "portal"],
+    "data": [
+        "security/ir.model.access.csv",
+        "security/portal_project_rules.xml",
+    ],
+    "demo": ["demo/demo_data.xml"],
+    "installable": True,
+    "application": False,
+    "auto_install": False,
+}
```
Don't worry about that

Mohajiro commented on project_internal_access_from_portal/models/project_project.py:
```diff
@@ -0,0 +1,14 @@
+class ProjectProject(models.Model):
+    _inherit = "project.project"
+
+    # Adding a new selection option to the privacy_visibility field
+    privacy_visibility = fields.Selection(
+        selection_add=[("portal_internal", "Invited internal/portal users")],
+        ondelete={"portal_internal": "set default"},
+        help="Show this project to invited internal users and portal users",
+    )
```
We don't need translation in the help section

Mohajiro commented on project_internal_access_from_portal/README.rst:
```diff
@@ -0,0 +1,96 @@
+Use Cases / Context
+===================
+
+Sometimes you need to provide access to a project to portal users. Even
+if this project privacy is set to the "Invited internal users"
+
```
it's okey

ivs-cetmix commented on project_internal_access_from_portal/models/project_project.py:
```diff
@@ -0,0 +1,14 @@
+    privacy_visibility = fields.Selection(
+        selection_add=[("portal_internal", "Invited internal/portal users")],
+        ondelete={"portal_internal": "set default"},
+        help="Show this project to invited internal users and portal users",
```
We are modifying an existing field, no need to add already present parameters.

ivs-cetmix commented on project_internal_access_from_portal/models/project_project.py:
```diff
@@ -0,0 +1,14 @@
+    # Adding a new selection option to the privacy_visibility field
+    privacy_visibility = fields.Selection(
+        selection_add=[("portal_internal", "Invited internal/portal users")],
+        ondelete={"portal_internal": "set default"},
```
Question: what is this key used for?

ivs-cetmix commented on project_internal_access_from_portal/security/ir.model.access.csv:
```diff
@@ -0,0 +1,3 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_project_portal,access.project.portal,project.model_project_project,base.group_portal,1,0,0,0
```
Access rights for the portal user are already [there](https://github.com/odoo/odoo/blob/61d3f8460eaf4217d258bab60f1da8d3281accab/addons/project/security/ir.model.access.csv#L17). No need to add them again.

ivs-cetmix commented on project_internal_access_from_portal/security/portal_project_rules.xml:
```diff
@@ -0,0 +1,28 @@
+    <record id="rule_portal_read_internal_projects" model="ir.rule">
+      <field name="name">Portal: read internal projects</field>
+      <field name="model_id" ref="project.model_project_project" />
+      <field name="domain_force">[('privacy_visibility','=','portal_internal')]</field>
```
This rule will allow portal users to see any project with the privacy set to "Portal/internal".  Which means that user Bob will see projects of the user Jane. Is that what we want to achieve?

ivs-cetmix commented on project_internal_access_from_portal/security/portal_project_rules.xml:
```diff
@@ -0,0 +1,28 @@
+      <field
+            name="domain_force"
+        >[('project_id.privacy_visibility','=','portal_internal')]</field>
+      <field name="groups" eval="[(4, ref('base.group_portal'))]" />
```
Same here

ivs-cetmix commented on project_internal_access_from_portal/demo/demo_data.xml:
```diff
@@ -0,0 +1,32 @@
+  <record id="demo_portal_internal_project" model="project.project">
+    <field name="name">Demo Internal/Portal Project</field>
+    <field name="privacy_visibility">portal_internal</field>
+    <field name="message_follower_ids" eval="[(4, ref('demo_portal_user_partner'))]" />
```
Please check which records are in the `message_follower_ids` field before using it.

ivs-cetmix reviewed (changes requested):
Please use existing source code as a reference. Open it and check how the same or similar things are already done in the module which you are extending.

OCA-git-bot committed:
Merge PR #1452 into 16.0

Signed-off-by pedrobaeza

Commit:
[UPD] Update project_milestone_status.pot

OCA-git-bot committed:
[BOT] post-merge updates

ivs-cetmix commented on project_internal_access_from_portal/tests/test_portal_internal_access.py:
```diff
@@ -0,0 +1,72 @@
+    - Portal user cannot write/create/unlink projects and tasks
+    """
+
+    def setUp(self):
```
This is Odoo 16.0, so you should use setUpClass instead.

ivs-cetmix reviewed (changes requested):
Hint: you can use https://search.odooism.com to search odoo code for getting some examples.

ivs-cetmix reviewed (approved):
Code review LGTM

victoralmau committed:
[FIX] project_stock: Generate analytical lines in both use cases (tasks and pickings)

In addition to generating the analytical lines at the end of the task, they must also be generated if the pickings is done.

Fixes https://github.com/OCA/project/issues/1492

OCA-git-bot committed:
Merge PR #1495 into 16.0

Signed-off-by pedrobaeza

OCA-git-bot committed:
[BOT] post-merge updates

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (2 of 2 strings)

Translation: project-16.0/project-16.0-project_task_parent_completion_blocking
Translate-URL: https://translation.odoo-community.org/projects/project-16-0/project-16-0-project_task_parent_completion_blocking/it/

mymage committed:
Added translation using Weblate (Italian)

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (8 of 8 strings)

Translation: project-16.0/project-16.0-project_milestone_status
Translate-URL: https://translation.odoo-community.org/projects/project-16-0/project-16-0-project_milestone_status/it/

CetmixGitDrone commented:
This PR has the `approved` label and has been created more than 5 days ago. It should therefore be ready to merge by a maintainer (or a PSC member if the concerned addon has no declared maintainer). 🤖
